src/idl_gen_cpp.cpp

```diff
@@ -2237,15 +2237,26 @@ class CppGenerator : public BaseGenerator {
       code_.SetValue("FIELD_VALUE",
                      GenUnderlyingCast(field, false, "_" + Name(field)));
 
-      code_ +=
-          "  bool mutate_{{FIELD_NAME}}({{FIELD_TYPE}} "
-          "_{{FIELD_NAME}}) {";
       if (false == field.IsScalarOptional()) {
         code_.SetValue("DEFAULT_VALUE", GenDefaultConstant(field));
+        code_.SetValue("FIELD_DEFAULT_VALUE", GenUnderlyingCast(field, true, GenDefaultConstant(field)));
+
+        // GenUnderlyingCast for a bool field generates 0 != 0
+        // So the type has to be checked and the appropriate default chosen
+        if (IsBool(field.value.type.base_type)) {
+          code_ += "  bool mutate_{{FIELD_NAME}}({{FIELD_TYPE}} "
```

Comment:
most of this line is still shared between the 3 cases, so can easily be outside the first `if` like before? Only duplicate what is actually different?

Reply:
Thank you! I see exactly what you mean, good catch.